### PR TITLE
Arachne 2372 code files allows duplicate

### PR DESCRIPTION
**Pull request by aklochkova**

### CODE REVIEW

File: src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAnalysisController.java [hunk 1]
```diff
@@ -444,30 +460,56 @@ public JsonResult updateCommonEntityInAnalysis(@PathVariable("analysisId") Long
...
+                try {
+                    createdFile = analysisService.saveFile(uploadFile.getFile(), user, analysis, uploadFile.getLabel(), uploadFile.getExecutable(), null);
+                } catch (IOException e) {
+                    LOGGER.error("Failed to save file", e);
+                } catch (AlreadyExistException e) {
+                    LOGGER.error("Failed to save file", e);
+                    result.setErrorCode(ALREADY_EXIST.getCode());
+                    errorFileMessages.add(e.getMessage());
+                }
+                createdFiles.add(conversionService.convert(createdFile, AnalysisFileDTO.class));
```
Comment (pavgra) [inline src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAnalysisController.java]: Shouldn't this be inside try block?

File: src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAnalysisController.java [hunk 2]
```diff
@@ -444,30 +460,56 @@ public JsonResult updateCommonEntityInAnalysis(@PathVariable("analysisId") Long
...
-            createdFiles.add(conversionService.convert(createdFile, AnalysisFileDTO.class));
-        } else {
-            if (StringUtils.hasText(uploadFileDTO.getLink())) {
-                AnalysisFile createdFile = analysisService.saveFile(uploadFileDTO.getLink(), user, analysis, uploadFileDTO.getLabel(), uploadFileDTO.getExecutable());
+        uploadFilesLinks.getLinks().forEach(uploadLink -> {
+            if (StringUtils.hasText(uploadLink.getLink())) {
+                AnalysisFile createdFile = null;
+                try {
+                    createdFile = analysisService.saveFile(uploadLink.getLink(), user, analysis, uploadLink.getLabel(), uploadLink.getExecutable());
+                } catch (IOException e) {
```
Comment (pavgra) [inline src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAnalysisController.java]: Looks very repetitive. Cannot we either move exception handling into some wrapper (so pass "saveFile" as lambda to a function, which wraps it with the error handlers) or use single forEach with internal logic for selecting which method to use (e.g. `if (StringUtils.hasText(uploadLink.getLink())) then uploadLink else uploadFile`)

File: src/main/java/com/odysseusinc/arachne/portal/api/v1/dto/converters/MultipartFileToUploadFileDTOConverter.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package com.odysseusinc.arachne.portal.api.v1.dto.converters;
```
Comment (agackovka) [inline src/main/java/com/odysseusinc/arachne/portal/api/v1/dto/converters/MultipartFileToUploadFileDTOConverter.java]: license header

File: src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAnalysisController.java [hunk 4]
```diff
@@ -444,34 +461,50 @@ public JsonResult updateCommonEntityInAnalysis(@PathVariable("analysisId") Long
...
-            if (StringUtils.hasText(uploadFileDTO.getLink())) {
-                AnalysisFile createdFile = analysisService.saveFile(uploadFileDTO.getLink(), user, analysis, uploadFileDTO.getLabel(), uploadFileDTO.getExecutable());
-                createdFiles.add(conversionService.convert(createdFile, AnalysisFileDTO.class));
-            } else {
-                result.setErrorCode(VALIDATION_ERROR.getCode());
-            }
-
+        Map<String, Object> validationErrors = new HashMap<>();
+        List<String> errorFileMessages = new ArrayList<>();
+        List<UploadFileDTO> files = uploadFilesLinks.getFiles().size() > 0 ? uploadFilesLinks.getFiles() : uploadFilesLinks.getLinks();
```
Comment (pavgra) [inline src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAnalysisController.java]: Why not concat?

File: src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAnalysisController.java [hunk 5]
```diff
@@ -444,34 +461,50 @@ public JsonResult updateCommonEntityInAnalysis(@PathVariable("analysisId") Long
...
+        saveFiles(files, user, analysis, result, createdFiles, errorFileMessages);
+        if (result.getErrorCode().equals(ALREADY_EXIST.getCode())) {
```
Comment (pavgra) [inline src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAnalysisController.java]: Couldn't we put this into the `saveFiles` function and then we could return `result` from there w/o changes by ref

File: src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAnalysisController.java (hunk 4 above)
Comment (aklochkova) [reply src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAnalysisController.java]: For what? We can have only either files collection or links but not both in the same time.

Comment (pavgra) [reply src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAnalysisController.java]: @aklochkova , although I see that there is no such use case currently, I also do not see a restriction in the DTO, which would prohibit presence of both

File: src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAnalysisController.java [hunk 6]
```diff
@@ -444,30 +462,47 @@ public JsonResult updateCommonEntityInAnalysis(@PathVariable("analysisId") Long
...
+                LOGGER.error("Failed to save file", e);
+                result.setErrorCode(ALREADY_EXIST.getCode());
+                errorFileMessages.add(e.getMessage());
             }
+            createdFiles.add(conversionService.convert(createdFile, AnalysisFileDTO.class));
+        });
 
+        if (result.getErrorCode().equals(ALREADY_EXIST.getCode())) {
+            validationErrors.put("file", errorFileMessages.toArray(new String[0]));
+            result.setValidatorErrors(validationErrors);
```
Comment (pavgra) [inline src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAnalysisController.java]: `ImmutableMap.of("file", errorFileMessages.toArray(new String[0]));`

and we do not need `validationErrors` var

**Pull request merged**